File: 0981-time-based-key-value-store/0981-time-based-key-value-store.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class TimeMap {

	private Map<String, ArrayList<Object[]>> store;
	
    public TimeMap() {
    	store = new HashMap<String, ArrayList<Object[]>>();
    }
    
    public void set(String key, String value, int timestamp) {
        
    	if(!store.containsKey(key)) {
    		store.put(key, new ArrayList<Object[]>());
    		store.get(key).add(new Object[] {timestamp, value});
    	}
    	else {
    		store.get(key).add(new Object[] {timestamp, value});
    	}
    }
    
    public String get(String key, int timestamp) {
    	
    	if(!store.containsKey(key)) {
    		return "";
    	}
    	
    	List<Object[]> valuesList = store.get(key);
    	
    	int left = 0;
    	int right = valuesList.size() - 1;
    	
    	while(left <= right) {
    		
    		int mid = (left + right) / 2 ;
    		
    		int mid_timestamp = (int)valuesList.get(mid)[0];
    		String mid_value = String.valueOf(valuesList.get(mid)[1]);
    		
    		if(mid_timestamp == timestamp) {
    			return mid_value;
    		}
    		else if (mid_timestamp < timestamp){
    			left = mid + 1;
    		}
    		else {
    			right = mid - 1;
    		}
    	}	
		if(right >= 0) {
			return String.valueOf(valuesList.get(right)[1]);
		}
		return "";
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.store = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.store:
            self.store[key] = []
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        
        timestamp_values = self.store[key]
        
        left = 0
        right = len(timestamp_values) - 1

        while left <= right:
            mid = (left + right) // 2

            mid_timestamp, mid_value = timestamp_values[mid]

            if(mid_timestamp == timestamp):
                return mid_value
            
            elif(mid_timestamp < timestamp):
                left = mid + 1
            else:
                right = mid - 1
        
        if right >= 0:
            return timestamp_values[right][1]
        
        return ""

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)